crates/accelerate/src/target_transpiler/mod.rs

```diff
@@ -0,0 +1,1285 @@
+            .unwrap()
+            .extract::<Option<Vec<String>>>()?;
+        self.non_global_strict_basis = state
+            .get_item("non_global_strict_basis")?
+            .unwrap()
+            .extract::<Option<Vec<String>>>()?;
+        Ok(())
+    }
+}
+
+// Rust native methods
+impl Target {
+    /// Returns an iterator over all the instructions present in the `Target`
+    /// as pair of `&TargetOperation` and `Option<&Qargs>`.
+    pub fn instructions(&self) -> impl Iterator<Item = (&TargetOperation, Option<&Qargs>)> {
+        self.gate_map.iter().flat_map(move |(op, props_map)| {
+            props_map
+                .keys()
+                .map(move |qargs| (&self._gate_name_map[op], qargs.as_ref()))
+        })
+    }
+    /// Returns an iterator over the operation names in the target.
+    pub fn operation_names(&self) -> impl Iterator<Item = &String> {
+        self.gate_map.keys()
+    }
+
+    /// Get the operation objects in the target.
+    pub fn operations(&self) -> impl Iterator<Item = &TargetOperation> {
+        return self._gate_name_map.values();
+    }
+
+    /// Get an iterator over the indices of all physical qubits of the target
+    pub fn physical_qubits(&self) -> impl Iterator<Item = usize> {
+        0..self.num_qubits.unwrap_or_default()
+    }
+
+    /// Generate non global operations if missing
+    fn generate_non_global_op_names(&mut self, strict_direction: bool) -> &[String] {
+        let mut search_set: HashSet<Qargs, RandomState> = HashSet::default();
+        if strict_direction {
+            // Build search set
+            for qarg_key in self.qarg_gate_map.keys().flatten().cloned() {
+                search_set.insert(qarg_key);
+            }
+        } else {
+            for qarg_key in self.qarg_gate_map.keys().flatten() {
+                if qarg_key.len() != 1 {
+                    let mut vec = qarg_key.clone();
+                    vec.sort_unstable();
+                    search_set.insert(vec);
+                }
+            }
+        }
+        let mut incomplete_basis_gates: Vec<String> = vec![];
+        let mut size_dict: IndexMap<usize, usize, RandomState> = IndexMap::default();
+        *size_dict
+            .entry(1)
+            .or_insert(self.num_qubits.unwrap_or_default()) = self.num_qubits.unwrap_or_default();
+        for qarg in &search_set {
+            if qarg.len() == 1 {
+                continue;
+            }
+            *size_dict.entry(qarg.len()).or_insert(0) += 1;
+        }
+        for (inst, qargs_props) in self.gate_map.iter() {
+            let mut qarg_len = qargs_props.len();
+            let qargs_keys: IndexSet<&Option<Qargs>> = qargs_props.keys().collect();
+            let qarg_sample = qargs_keys.iter().next().cloned();
+            if let Some(qarg_sample) = qarg_sample {
+                if !strict_direction {
+                    let mut qarg_set: HashSet<SmallVec<[PhysicalQubit; 2]>, RandomState> =
```

Comment:
I might rename this `deduplicated_qargs` because it wasn't entirely clear what the purpose was here. I assume it's for deduplication because otherwise `qarg_len` will always be `qargs_keys.len()`. That being said I wonder if we can move this definition outside the loop and reuse a single allocation for each instruction.

Reply:
Will rename, and I think it is already a single allocation for each instruction as these deduplicated `qargs` are those that are part of the instruction's property map.